<?xml version="1.0"?>
<robot name="mars_rover_iitb" xmlns:xacro="http://ros.org/wiki/xacro">

  <!-- Included URDF/XACRO Files -->
  <!-- <xacro:include filename="$(find husky_description)/urdf/decorations.urdf.xacro" /> -->
  <xacro:include filename="$(find msi_2k16_17_urdf)/urdf/wheel.urdf.xacro" />
  <xacro:include filename="$(find msi_2k16_17_urdf)/urdf/arm.urdf" />

  <material name="chassis">
    <color rgba="0.67 0.78 0.85 1" />
  </material>

  <property name="M_PI" value="3.14159"/>

  <!-- Base Size -->
  <property name="base_x_size" value="1.1" />
  <property name="base_y_size" value="0.59" />
  <property name="base_z_size" value="0.35" />

  <property name="rocker_wheel_x" value="0.4947" />
  <property name="rocker_wheel_y" value="0.1083" />
  <property name="rocker_wheel_z" value="-0.2738" />

  <property name="bogie_wheel_x" value="0.2348" />
  <property name="bogie_wheel_y" value="0.1146" />
  <property name="bogie_wheel_z" value="-0.1932" />

  <!-- Wheel Properties -->

  <property name="wheel_L" value="0.18" />
  <property name="wheel_R" value="0.17" />
  <property name="wheel_length" value="0.18" />
  <property name="wheel_radius" value="0.17" />


  <!-- Wheel Mounting Positions -->
  <xacro:property name="wheelbase" value="1.08" />
  <xacro:property name="track" value="1.0" />
  <xacro:property name="wheel_vertical_offset" value="${-base_z_size/2}" />


    <!-- Base link is the center of the robot's bottom plate -->
  <link name="base_link">
    <visual>
      <geometry>
         <mesh filename="package://msi_2k16_17_urdf/meshes/chassis.stl" scale="0.01 0.01 0.01"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 ${-M_PI/2}"/>
      <material name="chassis"/>
    </visual>
  </link>

  <gazebo reference="base_link">
    <material> Gazebo/Grey</material>
  </gazebo>

  <link name="rocker_R">
     <visual>
      <geometry>
         <mesh filename="package://msi_2k16_17_urdf/meshes/rocker_MIR.stl" scale="0.01 0.01 0.01"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>

  <joint name="rocker_R_to_chassis" type="revolute">
      <origin xyz=" 0 ${base_y_size/2 + 0.04} 0" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="rocker_R"/>

<limit effort="20" lower="-0.17" upper="0.17" velocity="1"/>
      <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>

  <link name="rocker_L">
     <visual>
      <geometry>
         <mesh filename="package://msi_2k16_17_urdf/meshes/rocker.stl" scale="0.01 0.01 0.01"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>

  <joint name="rocker_L_to_chassis" type="revolute">
      <origin xyz=" 0 ${-base_y_size/2 - 0.02} 0" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="rocker_L"/>

<limit effort="20" lower="-0.17" upper="0.17" velocity="1"/>
      <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>


  <link name="bogie_R">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/bogie.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>

  <joint name="bogie_R_to_rocker" type="revolute">
      <origin xyz="-0.202 0.0200 -0.1434" rpy="0 0 0"/>
      <parent link="rocker_R"/>
      <child link="bogie_R"/>

<limit effort="20" lower="-0.17" upper="0.17" velocity="1"/>
      <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>

  <link name="bogie_L">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/bogie.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>

  <link name="steer_wheel_fL">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/steer_wheel_l.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>
  <link name="steer_wheel_fR">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/steer_wheel_r.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>
  <link name="steer_wheel_bL">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/steer_wheel_l.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>
  <link name="steer_wheel_bR">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/steer_wheel_r.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>

  <link name="drive_wheel_R">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/drive_wheel_l.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>

  <link name="drive_wheel_L">
    <visual>
     <geometry>
       <mesh filename="package://msi_2k16_17_urdf/meshes/drive_wheel_r.stl" scale="0.01 0.01 0.01"/>
     </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="chassis"/>
    </visual>
    <inertial>
      <origin xyz="0 0 1" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/>
    </inertial>
  </link>
  <joint name="bogie_L_to_rocker" type="revolute">
      <origin xyz="-0.202 -0.044 -0.1434 " rpy="0 0 0"/>
      <parent link="rocker_L"/>
      <child link="bogie_L"/>


<limit effort="20" lower="-0.17" upper="0.17" velocity="1"/>
      <axis xyz="0 1 0" rpy="0 0 0" />
  </joint>



    <!-- Base footprint is on the ground under the robot -->
    <link name="base_footprint"/>

    <joint name="base_footprint_joint" type="fixed">
      <origin xyz="0 0 ${wheel_vertical_offset - wheel_R}" rpy="0 0 0" />
      <parent link="base_link" />
      <child link="base_footprint" />
    </joint>

    <!-- Interial link stores the robot's inertial information -->
    <link name="inertial_link">
      <inertial>
        <mass value="33.455" />
        <origin xyz="-0.08748 -0.00085 0.09947" />
        <inertia ixx="0.6022" ixy="-0.02364" ixz="-0.1197" iyy="1.7386" iyz="-0.001544" izz="2.0296" />
      </inertial>
    </link>

    <joint name="inertial_joint" type="fixed">
      <origin xyz="0 0 0" rpy="0 0 0" />
      <parent link="base_link" />
      <child link="inertial_link" />
    </joint>

    <!-- IMU Link is the standard mounting position for the UM6 IMU.-->
    <!-- Can be modified with environment variables in /etc/ros/setup.bash -->
    <link name="imu_link"/>
    <joint name="imu_joint" type="fixed">
      <origin xyz="0.19 0 0.149" rpy="0 -1.5708 3.1416" />
      <parent link="base_link" />
      <child link="imu_link" />
    </joint>
    <gazebo reference="imu_link">
    </gazebo>

    <!-- wheel macros -->
     <!-- STATIC joints for steer assembly -->
      <joint name="bogie_to_steer_L" type="fixed">          
  			<origin xyz="${-bogie_wheel_x} 0 ${bogie_wheel_z+0.2}" rpy="0 0 0" />
            <parent link="bogie_L"/>
            <child link="steer_wheel_fL"/>
        </joint>

      <joint name="bogie_to_steer_R" type="fixed">          
        <origin xyz="${-bogie_wheel_x} 0 ${bogie_wheel_z+0.2}" rpy="0 0 0" />
            <parent link="bogie_R"/>
            <child link="steer_wheel_fR"/>
        </joint>

        <joint name="rocker_to_steer_R" type="fixed">          
              <origin xyz="${rocker_wheel_x} 0 ${rocker_wheel_z+0.2}" rpy="0 0 0" />
              <parent link="rocker_R"/>
              <child link="steer_wheel_bR"/>
        </joint>

        <joint name="rocker_to_steer_L" type="fixed">          
            <origin xyz="${rocker_wheel_x} 0 ${rocker_wheel_z+0.2}" rpy="0 0 0" />
            <parent link="rocker_L"/>
            <child link="steer_wheel_bL"/>
      </joint>

   <joint name="bogie_to_drive_L" type="fixed">          
        <origin xyz="${bogie_wheel_x} 0.1 ${bogie_wheel_z+0.05}" rpy="0 0 0" />
            <parent link="bogie_L"/>
            <child link="drive_wheel_L"/>
        </joint>

      <joint name="bogie_to_drive_R" type="fixed">          
        <origin xyz="${bogie_wheel_x} -0.1 ${bogie_wheel_z+0.05}" rpy="0 0 0" />
            <parent link="bogie_R"/>
            <child link="drive_wheel_R"/>
        </joint>


      <!-- Join arm -->


    <xacro:husky_wheel wheel_prefix="front_L" parent="steer_wheel_fL">
  <origin xyz="0.04 -0.13 -0.22" rpy="0 0 0" />
    </xacro:husky_wheel>

   <xacro:husky_wheel wheel_prefix="front_R" parent="steer_wheel_fR">
  <origin xyz="-0.04 0.13 -0.22" rpy="0 0 0" />
    </xacro:husky_wheel>
  
   <xacro:husky_wheel wheel_prefix="rear_L" parent="steer_wheel_bL">
  <origin xyz="0.04 -0.13 -0.22" rpy="0 0 0" />
    </xacro:husky_wheel>

   <xacro:husky_wheel wheel_prefix="rear_R" parent="steer_wheel_bR">
  <origin xyz="-0.04 0.13 -0.22" rpy="0 0 0" />
    </xacro:husky_wheel>

  <!-- Center Wheel Handled Separately; Only Drive, no Steer -->
   <link name="center_wheel_link_L">
      <inertial>
        <mass value="5" />
        <origin xyz="0 0 0" />
        <inertia  ixx="0.02467" ixy="0" ixz="0" iyy="0.02467" iyz="0" izz="0.04411" />
      </inertial>
      <visual>
        <origin xyz="0 ${-wheel_L/2} 0" rpy="0 0 ${M_PI/2}" />
        <geometry>
          <mesh filename="package://msi_2k16_17_urdf/meshes/tyre.stl" scale="0.01 0.01 0.01" />
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="${M_PI/2} 0 0" />
        <geometry>
          <cylinder length="${wheel_length}" radius="${wheel_radius}" />
        </geometry>
      </collision>
    </link>

    <link name="center_wheel_link_R">
      <inertial>
        <mass value="5" />
        <origin xyz="0 0 0" />
        <inertia  ixx="0.02467" ixy="0" ixz="0" iyy="0.02467" iyz="0" izz="0.04411" />
      </inertial>
      <visual>
        <origin xyz="0 ${-wheel_L/2} 0" rpy="0 0 ${M_PI/2}" />
        <geometry>
          <mesh filename="package://msi_2k16_17_urdf/meshes/tyre.stl" scale="0.01 0.01 0.01" />
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="${M_PI/2} 0 0" />
        <geometry>
          <cylinder length="${wheel_length}" radius="${wheel_radius}" />
        </geometry>
      </collision>
    </link>



    <gazebo reference="center_wheel_link_R">
      <mu1 value="1.0"/>
      <mu2 value="1.0"/>
      <kp value="10000000.0" />
      <kd value="1.0" />
      <fdir1 value="1 0 0"/>
        <material> Gazebo/FlatBlack</material>
    </gazebo>


    <gazebo reference="center_wheel_link_L">
      <mu1 value="1.0"/>
      <mu2 value="1.0"/>
      <kp value="10000000.0" />
      <kd value="1.0" />
      <fdir1 value="1 0 0"/>
        <material> Gazebo/FlatBlack</material>
    </gazebo>

    <joint name="center_wheel_drive_L" type="continuous">
      <parent link="drive_wheel_L"/>
      <child link="center_wheel_link_L"/>
      <origin xyz="0 -0.23 0" rpy="0 0 0" />
      <axis xyz="0 1 0" rpy="0 0 0" /> 
    </joint>

    <transmission name="center_wheel_L_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="center_wheel_motor_L">
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="center_wheel_drive_L">
        <hardwareInterface>VelocityJointInterface</hardwareInterface>
      </joint>
    </transmission>

    <joint name="center_wheel_drive_R" type="continuous">
      <parent link="drive_wheel_R"/>
      <child link="center_wheel_link_R"/>
      <origin xyz="0 0.23 0" rpy="0 0 0" />
      <axis xyz="0 1 0" rpy="0 0 0" />
    </joint>

    <transmission name="center_wheel_R_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="center_wheel_motor_R">
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="center_wheel_drive_R">
        <hardwareInterface>VelocityJointInterface</hardwareInterface>
      </joint>
    </transmission>



    <!-- Decorations MAcro -->
    <!-- <xacro:husky_decorate /> -->



      <xacro:property name="reflect" value="-1" />

      <!-- Differential Suspension -->
      <link name="differential" >
         <visual>
          <geometry>
             <mesh filename="package://msi_2k16_17_urdf/meshes/differential.stl" scale="0.01 0.01 0.01"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <material name="chassis"/>
        </visual>
        <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="1"/>
          <inertia
            ixx="1.0" ixy="0.0" ixz="0.0"
            iyy="1.0" iyz="0.0"
            izz="1.0"/>
        </inertial>
      </link>

      <!-- STATIC joints for differential -->
      <joint name="diff_to_chassis" type="fixed">
                  <origin xyz="${base_x_size/2-0.12} 0 0" rpy="0 0 0"/>
                  <parent link="base_link"/>
                    <child link="differential"/>
                  </joint>

      <!-- Join arm -->


     <joint name="jaco_arm_base" type="fixed">
        <parent link="base_link" />
        <child link="base_link1" />
        <origin xyz=" -0.2 0 0" rpy="0 0 ${M_PI}" />
      </joint>



</robot>
